TESTS/mbedtls/multi/main.cpp

```diff
@@ -0,0 +1,146 @@
+    printf("\nawaited result  : 50EA825D9684F4229CA29F1FEC511593E281E46A140D81E0005F8F688669A06C\n"); // for  abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq
+  
+    printf("\nend of test sha256\n");
+  TEST_ASSERT_EQUAL_UINT8_ARRAY(outsum, test_sum,32);
+}
+
+/* Tests that treating 2 sha256 objects in // does not impact the result */
+void test_case_sha256_multi() {
+    const unsigned char test_buf[] = {"abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopqabcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopqabcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq"};
+    // sha256_output_values for abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq
+    const unsigned char test_sum1[] =
+    { 0x24, 0x8D, 0x6A, 0x61, 0xD2, 0x06, 0x38, 0xB8,
+      0xE5, 0xC0, 0x26, 0x93, 0x0C, 0x3E, 0x60, 0x39,
+      0xA3, 0x3C, 0xE4, 0x59, 0x64, 0xFF, 0x21, 0x67,
+      0xF6, 0xEC, 0xED, 0xD4, 0x19, 0xDB, 0x06, 0xC1 };
+    // sha256_output_values for 3*abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq
+    const unsigned char test_sum2[] =
+    { 0x50, 0xEA, 0x82, 0x5D, 0x96, 0x84, 0xF4, 0x22,
+      0x9C, 0xA2, 0x9F, 0x1F, 0xEC, 0x51, 0x15, 0x93,
+      0xE2, 0x81, 0xE4, 0x6A, 0x14, 0x0D, 0x81, 0xE0,
+      0x00, 0x5F, 0x8F, 0x68, 0x86, 0x69, 0xA0, 0x6C};
+    unsigned char outsum1[32], outsum2[32];
+    int i;
+
+    mbedtls_sha256_context ctx1;
+    mbedtls_sha256_context ctx2;
+    printf("test sha256_multi\n");
+    //Init both contexts
+    mbedtls_sha256_init( &ctx1 );
+    mbedtls_sha256_init( &ctx2 );
+    //Start both contexts
+    mbedtls_sha256_starts( &ctx1, 0);
+    mbedtls_sha256_starts( &ctx2, 0);
+
+    printf("upd ctx1\n");
+    mbedtls_sha256_update( &ctx1, test_buf, 56 );
+    printf("upd ctx2\n");
+    mbedtls_sha256_update( &ctx2, test_buf, 66 );
+    printf("finish ctx1\n");
+    mbedtls_sha256_finish( &ctx1, outsum1 );
+    printf("upd ctx2\n");
+    mbedtls_sha256_update( &ctx2, test_buf+66, 46 );
+    printf("free ctx1\n");
+    mbedtls_sha256_free( &ctx1 );
+    printf("upd ctx2\n");
+    mbedtls_sha256_update( &ctx2, test_buf+112, 56 );
+    printf("finish ctx2\n");
+    mbedtls_sha256_finish( &ctx2, outsum2 );
+    printf("free ctx2\n");
+    mbedtls_sha256_free( &ctx2 );
+  
+    printf("\nreceived result ctx1 : ");
+    for (i=0;i<32;i++) { printf("%02X",outsum1[i]);}
+    printf("\nawaited result       : 248D6A61D20638B8E5C026930C3E6039A33CE45964FF216F6ECEDD19DB06C1\n"); // for  abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq
+    printf("\nreceived result ctx2 : ");
+    for (i=0;i<32;i++) { printf("%02X",outsum2[i]);}
+    printf("\nawaited result       : 50EA825D9684F4229CA29F1FEC511593E281E46A140D81E0005F8F688669A06C\n"); // for  3*abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq
+  
+    printf("\nend of test sha256\n");
+    TEST_ASSERT_EQUAL_UINT8_ARRAY(outsum1, test_sum1,32);
+    TEST_ASSERT_EQUAL_UINT8_ARRAY(outsum2, test_sum2,32);
+}
+#endif //MBEDTLS_SHA256_C
+
```

Comment:
Nice tests. I suggest testing {{mbedtls_sha256_clone}} as well. You can follow the same steps as in `md_text_multi` in [the mbed TLS library unit tests](https://github.com/ARMmbed/mbedtls/blob/development/tests/suites/test_suite_md.function). Or you can clone {{ctx2}} to a new {{ctx3}} after the first {{update}} on {{ctx2}}, then update {{ctx3}} with a different last part from {{ctx2}}, and check that the result is correct for both 2 and 3.

By the way, you can use the interface in {{mbedtls/md.h}} with the {{mbedtls_md_XXX}} functions to write the function once and then be able to execute it for MD5, SHA-1 and SHA-256.